src/probe/probe.c

```diff
@@ -15,42 +31,1072 @@
+			return -EINVAL;
+		}
+		pbuf->r_ptr = pbuf->r_ptr + tail;
+	} else {
+		pbuf->r_ptr = pbuf->r_ptr + head;
+	}
+
+	/* subtract used bytes */
+	pbuf->avail -= bytes;
+
 	return 0;
 }
 
+/**
+ * \brief Generate probe data packet header, update timestamp, calc crc
+ *	  and copy data to probe buffer.
+ * \param[in] component buffer pointer.
+ * \param[in] data size.
+ * \param[in] audio format.
+ * \return 0 on success, error code otherwise.
+ */
+static int probe_gen_header(struct comp_buffer *buffer, uint32_t size, uint32_t format)
+{
+	struct probe_pdata *_probe = probe_get();
+	struct probe_data_packet *header;
+	uint64_t timestamp;
+	uint32_t crc;
+
+	header = &_probe->header;
+	timestamp = platform_timer_get(timer_get());
+
+	header->sync_word = PROBE_EXTRACT_SYNC_WORD;
+	header->buffer_id = buffer->id;
+	header->format = format;
+	header->timestamp_low = (uint32_t)timestamp;
+	header->timestamp_high = (uint32_t)(timestamp >> 32);
+	header->checksum = 0;
+	header->data_size_bytes = size;
+
+	/* calc crc to check validation by probe parse app */
+	crc = crc32(0, header, sizeof(*header));
+	header->checksum = crc;
+
+	dcache_writeback_region(header, sizeof(*header));
+
+	return copy_to_pbuffer(&_probe->ext_dma.dmapb, header,
+			       sizeof(struct probe_data_packet));
+}
+
+/**
+ * \brief Generate description of audio format for extraction probes.
+ * \param[in] frame_fmt.
+ * \param[in] sample rate.
+ * \param[in] channels num.
+ * \return format.
+ */
+static uint32_t probe_gen_format(uint32_t frame_fmt, uint32_t rate, uint32_t channels)
+{
+	uint32_t format = 0;
+	uint32_t sample_rate;
+	uint32_t valid_bytes;
+	uint32_t container_bytes;
+	uint32_t float_fmt = 0;
+
+	switch (frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		valid_bytes = 2;
+		container_bytes = 2;
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+		valid_bytes = 3;
+		container_bytes = 4;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		valid_bytes = 4;
+		container_bytes = 4;
+		break;
+	case SOF_IPC_FRAME_FLOAT:
+		valid_bytes = 4;
+		container_bytes = 4;
+		float_fmt = 1;
+		break;
+	default:
+		trace_probe_error("probe_gen_format() error: Invalid frame format specified = 0x%08x",
+				  frame_fmt);
+		assert(false);
+	}
+
+	switch (rate) {
+	case 8000:
+		sample_rate = 0;
+		break;
+	case 11025:
+		sample_rate = 1;
+		break;
+	case 12000:
+		sample_rate = 2;
+		break;
+	case 16000:
+		sample_rate = 3;
+		break;
+	case 22050:
+		sample_rate = 4;
+		break;
+	case 24000:
+		sample_rate = 5;
+		break;
+	case 32000:
+		sample_rate = 6;
+		break;
+	case 44100:
+		sample_rate = 7;
+		break;
+	case 48000:
+		sample_rate = 8;
+		break;
+	case 64000:
+		sample_rate = 9;
+		break;
+	case 88200:
+		sample_rate = 10;
+		break;
+	case 96000:
+		sample_rate = 11;
+		break;
+	case 128000:
+		sample_rate = 12;
+		break;
+	case 176400:
+		sample_rate = 13;
+		break;
+	case 192000:
+		sample_rate = 14;
+		break;
+	default:
+		sample_rate = 15;
```

Comment:
An enum would be nice.

Reply:
Enum for rates would be inconsistent with rest of code where we pass it just by value everywhere. However this switch should be chaned to some  some array like `supported_rates={8000,16000,...}` and then search for rate in it.

Reply:
Changed to array